File: LifeCycle/valid/lifecycleserver2/java/src/cdmw/validation/UserResourceServantProvider.java
/* ===================================================================== */
/*
 * This file is part of CARDAMOM (R) which is jointly developed by THALES 
 * and SELEX-SI. 
 * 
 * It is derivative work based on PERCO Copyright (C) THALES 2000-2003. 
 * All rights reserved.
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it under 
 * the terms of the GNU Library General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your 
 * option) any later version. 
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT 
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public 
 * License for more details. 
 * 
 * You should have received a copy of the GNU Library General 
 * Public License along with CARDAMOM; see the file COPYING. If not, write to 
 * the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
/* ===================================================================== */

package cdmw.validation;

import cdmw.lifecycle.ObjectDeactivator;
import cdmw.lifecycle.LifeCycleObjectBase;
import cdmw.resources.ApplicationImpl;
import cdmw.resources.DeviceImpl;
import cdmw.resources.ResourceServantProvider;

import com.thalesgroup.CF.ApplicationPOATie;
import com.thalesgroup.CF.DataType;
import com.thalesgroup.CF.InvalidResourceNumber;
import com.thalesgroup.CF.InvalidProperties;
import com.thalesgroup.CF.CreationError;
import com.thalesgroup.CF.CreateApplicationError;
import com.thalesgroup.CF.DevicePackage.UsageType;
import com.thalesgroup.CF.DevicePackage.InvalidLocation;

import org.omg.CORBA.ORB;
import org.omg.CORBA.SystemException;
import org.omg.CORBA.OBJECT_NOT_EXIST;
import org.omg.CORBA.BAD_OPERATION;
import org.omg.CORBA.BAD_PARAM;
import org.omg.PortableServer.POA;
import org.omg.PortableServer.Servant;
import org.omg.PortableServer.ForwardRequest;

/**
 * UserResourceServantProvider
 *
 */
public class UserResourceServantProvider extends ResourceServantProvider {

    public static final short DEV = 1;
    public static final short APP = 2;

    public static final String DEV_PREFIX = "DEV ";
    public static final String APP_PREFIX = "APP ";

    public static final String DEV_REP_ID
        = "IDL:thalesgroup.com/CF/Device:1.0";
    public static final String APP_REP_ID
        = "IDL:thalesgroup.com/CF/Application:1.0";

    // Properties Identifiers
    public static final String LABEL_ID = "LABEL";
    public static final String NAME_ID  = "NAME";

    /**
     * [Attribute description]
     */
    private ORB orb;

    /**
     * [Attribute description]
     */
    private POA poa;

    /**
     * [Attribute description]
     */
    private ObjectDeactivator objectDeactivator;

    /**
     * Active Object Map
     */
    private java.util.HashMap aom = new java.util.HashMap();

    public class ObjInfo {

        private short type;

        private Servant serv;

        public ObjInfo() {
            this.type = 0;
            this.serv = null;
        }

        public ObjInfo(short type, Servant serv) {
            this.type = type;
            this.serv = serv;
        }

        public short getType() {
            return this.type;
        }

        public Servant getServant() {
            return this.serv;
        }

    }

    /**
     * Constructor
     *
     * @param orb
     * @param poa
     * @param deactivator
     */
    public UserResourceServantProvider(ORB orb, POA poa,
        ObjectDeactivator deactivator) {

        this.orb = orb;
        this.poa = poa;
        this.objectDeactivator = deactivator;

    }

    /**
     * This method acts as a hand shaking between the Servant Provider
     * and the generated Factory implementation. It indicates whether
     * a Servant provided by this Servant Provider may incarnate multiple
     * objects (with ObjectIds provided by this Servant Provider)
     *
     * @return <B>true</B> if a servant may incarnate multiple objects,
     * <B>false</B> otherwise.
     */
    public boolean mayIncarnateMultipleObjects() {
        return false;
    }

    /**
     * Create a resource.
     *
     * @param resourceNumber
     * @param qualifiers
     *
     * @return Resource Object Id
     *
     * @exception InvalidResourceNumber
     * @exception InvalidProperties
     * @exception CreationError
     */
    public byte[] create_resource(short resourceNumber, DataType[] qualifiers)
        throws  InvalidResourceNumber,
                InvalidProperties,
                CreationError {

        if (resourceNumber == DEV) {
            if ( (qualifiers.length == 1) &&
                (qualifiers[0].id.equals(LABEL_ID)) ) {
                try {
                    String label = qualifiers[0].value.extract_string();
                    byte[] oid = create_device(label);
                    return oid;
                } catch (BAD_OPERATION bp) {
                    String[] messages = {"Wrong label value."};
                    throw new InvalidProperties(messages);
                } catch (InvalidLocation il) {
                    throw new CreationError();
                }
            } else {
                String[] messages = {"Label property is missing."};
                throw new InvalidProperties(messages);
            }
        } else if (resourceNumber == APP) {
            if ( (qualifiers.length == 1) &&
                (qualifiers[0].id.equals(NAME_ID)) ) {
                try {
                    String name = qualifiers[0].value.extract_string();
                    return create_application(name);
                } catch (BAD_OPERATION bp) {
                    String[] messages = {"Wrong label value."};
                    throw new InvalidProperties(messages);
                } catch (CreateApplicationError cae) {
                    throw new CreationError();
                }
            } else {
                String[] messages = {"Name property is missing."};
                throw new InvalidProperties(messages);
            }
        } else {
            throw new InvalidResourceNumber();
        }

    }

    /**
     * Create a device.
     *
     * @param label
     *
     * @return Device Object Id
     *
     * @exception InvalidLocation
     */
    public byte[] create_device(String label) throws InvalidLocation {

        try {
            String key = DEV_PREFIX + label;
            if ( aom.get(key)!=null ) throw new InvalidLocation();

            byte[] oid = key.getBytes();
            ObjInfo info = new ObjInfo(DEV,
                new DeviceImpl(poa, objectDeactivator, UsageType.IDLE, label));
            aom.put(key, info);
            System.out.println("DEVICE <" + key
                + "> created with LABEL='" + label + "'");
            return oid;
        } catch (Exception e) {
            throw new InvalidLocation();
        }

    }

    /**
     * Create an application.
     *
     * @param name
     *
     * @return Application Object Id
     *
     * @exception CreateApplicationError
     */
    public byte[] create_application(String name)
        throws CreateApplicationError {

        try {
            String key = APP_PREFIX + name;
            if (aom.get(key) != null) {
                System.out.println("Application already active");
                throw new CreateApplicationError();
            }

            byte[] oid = key.getBytes();
            ApplicationImpl applicationImpl
                = new ApplicationImpl(objectDeactivator, name);
            ObjInfo info = new ObjInfo(APP,
                new ApplicationPOATie(applicationImpl, poa));
            aom.put(key, info);
            System.out.println("APPLICATION <" + key +
                "> created with NAME='" + name + "'");
            return oid;
        } catch (Exception e) {
            throw new CreateApplicationError();
        }

    }

    /**
     * Creates a servant depending on its type
     *
     * @param oid Object Id to lookup for
     * @return A new servant
     * @exception ForwardRequest
     */
    public Servant lookupServant(byte[] oid) throws ForwardRequest {

        try  {
            String key = new String(oid);
            ObjInfo objInfo = (ObjInfo) aom.get(key);
            if (objInfo == null) {
                throw new OBJECT_NOT_EXIST();
            } else {
                System.out.println("Activating " + key + " ... ");
                return objInfo.getServant();
            }
        } catch (BAD_PARAM bp) { // Pb with ObjectId
            throw new OBJECT_NOT_EXIST(); // TODO Minor codes
        }

    }

    /**
     * Releases the servant passed in parameter
     *
     * @param oid
     * @param poa
     * @param servant
     */
    public void releaseServant(byte[] oid,POA poa,Servant servant) {

        try {
            LifeCycleObjectBase removable = (LifeCycleObjectBase) servant;
            StringBuffer echo = new StringBuffer();
            String key = new String(oid);
            echo.append("Deactivating " + key + " ... ");
            if (removable != null && removable.isRemoved(oid)) {
                cleanupObject(oid);
                aom.remove(key);
                echo.append("deleted.");
            }
            System.out.println(echo.toString());
        } catch (BAD_PARAM bp) { // A problem with oid
            // TODO throw an exception
        } catch(ClassCastException cce) {}

    }

    /**
     * Returns the repository id of the object with oid
     *
     * @param oid
     * @return The supported interface
     * @exception OBJECT_NOT_EXIST
     */
    public String getSupportedInterface(byte[] oid)
        throws OBJECT_NOT_EXIST {

        try {
            String name = new String(oid);
            ObjInfo objInfo = (ObjInfo) aom.get(name);
            if(objInfo == null) {
                throw new OBJECT_NOT_EXIST(); // TODO Minor codes
            } else {
                String result = new String();
                if (objInfo.getType() == DEV) {
                    result = DEV_REP_ID;
                } else {
                    result = APP_REP_ID;
                }
                return result;
            }
        } catch (BAD_PARAM bp) { // Pb with ObjectId
            // TODO throw an exception
        }
        return ""; // to silence the compiler


    }

}

File: LifeCycle/valid/lifecycleserver1/c++/src/UserResourceServantProvider.cpp
/* ===================================================================== */
/*
 * This file is part of CARDAMOM (R) which is jointly developed by THALES
 * and SELEX-SI. It is derivative work based on PERCO Copyright (C) THALES
 * 2000-2003. All rights reserved.
 * 
 * Copyright (C) THALES 2004-2005. All rights reserved
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Library General Public License as published
 * by the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public
 * License for more details.
 * 
 * You should have received a copy of the GNU Library General Public
 * License along with CARDAMOM; see the file COPYING. If not, write to the
 * Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
/* ===================================================================== */

#include "lifecycleserver1/UserResourceServantProvider.hpp"
#include "lifecycleserver1/Device_impl.hpp"
#include "lifecycleserver1/Application_impl.hpp"
#include "Foundation/common/Assert.hpp"
#include "lifecycleserver1/MacroDefs.hpp"

#include <sstream>

namespace {
    const CF::ResourceNumType DEV = 1;
    const CF::ResourceNumType APP = 2;
    
    const char* DEV_PREFIX = "DEV ";
    const char* APP_PREFIX = "APP ";
    
    const char* DEV_REP_ID = "IDL:thalesgroup.com/CF/Device:1.0";
    const char* APP_REP_ID = "IDL:thalesgroup.com/CF/Application:1.0";
    
    // Properties Identifiers
    const char* LABEL_ID = "LABEL";
    const char* NAME_ID  = "NAME";
};

namespace Cdmw
{
namespace Resources
{

/**
*Implementation notes: [if no pertinent write none ]
*<p>
*Portability issues: [if no pertinent write none ]
*<p>
*/

UserResourceServantProvider::~UserResourceServantProvider()
    throw()
{
    // Delete all servants
    ActiveObjectMap::iterator info_pos = m_aom.begin();

    while (info_pos != m_aom.end())
    {
        PortableServer::Servant serv = (info_pos->second).m_serv;
        CDMW_ASSERT(serv != 0);
        serv->_remove_ref();
	info_pos++;
    }
}

// UserResourceServantProvider::UserResourceServantProvider()
//     throw()
//   : m_orb(CORBA::ORB::_nil()),
//     m_poa(PortableServer::POA::_nil())
// {
// }

UserResourceServantProvider::UserResourceServantProvider(CORBA::ORB_ptr          orb,
                                                         PortableServer::POA_ptr poa,
                                                         const Cdmw::LifeCycle::ObjectDeactivator & deactivator)
  throw(CORBA::SystemException)
  : m_orb(CORBA::ORB::_duplicate(orb)),
    m_poa(PortableServer::POA::_duplicate(poa)),
    m_object_deactivator(deactivator)
{
}

PortableServer::ObjectId * 
UserResourceServantProvider::create_resource(CF::ResourceNumType   resource_number,
                                             const CF::Properties& qualifiers)
    throw(CF::InvalidResourceNumber,
          CF::InvalidProperties,
          CF::CreationError)
{ 
  if (resource_number == DEV) {
      if ((qualifiers.length() == 1) &&
          (strcmp(qualifiers[0].id.in(),LABEL_ID) == 0)) {
          try {
              const char* label;
              qualifiers[0].value >>= label;
              PortableServer::ObjectId_var oid
                  = create_device(label);
              return oid._retn();
          } catch (const CORBA::DATA_CONVERSION & ) {
              CF::InvalidProperties ex;
              ex.error_messages.length(1);
              ex.error_messages[0] = CORBA::string_dup("Wrong label value.");
              throw  ex;
          } catch (const CF::Device::InvalidLocation & ) {
              throw  CF::CreationError();
          } 
      } 
      else {
          CF::InvalidProperties ex;
          ex.error_messages.length(1);
          ex.error_messages[0] = CORBA::string_dup("Label property is missing.");
          throw  ex;
      }
  } else if (resource_number == APP) {
      if ((qualifiers.length() == 1) &&
          (strcmp(qualifiers[0].id.in(),NAME_ID) == 0)) {
          try {
              const char* name;
              qualifiers[0].value >>= name;
              return create_application(name);
          } catch (const CORBA::DATA_CONVERSION & ) {
              CF::InvalidProperties ex;
              ex.error_messages.length(1);
              ex.error_messages[0] = CORBA::string_dup("Wrong name value.");
              throw  ex;
          } 
      }
      else {
          CF::InvalidProperties ex;
          ex.error_messages.length(1);
          ex.error_messages[0] = CORBA::string_dup("Name property is missing.");
          throw  ex;
      }
  } else
    throw CF::InvalidResourceNumber();
}

PortableServer::ObjectId * 
UserResourceServantProvider::create_device(const char* label)
    throw(CF::Device::InvalidLocation)
{
    try {
        std::string key(DEV_PREFIX);
        key += label;
        ActiveObjectMap::iterator serv_pos = m_aom.find(key);
        if (serv_pos  != m_aom.end())
            throw CF::Device::InvalidLocation();

        PortableServer::ObjectId_var oid
            = PortableServer::string_to_ObjectId(key.c_str());

        ObjInfo info(DEV,new Device_impl(m_poa.in(),m_object_deactivator,CF::Device::IDLE,label));
        m_aom[key] = info;
        ECHO_CDMW << "DEVICE <" << key << "> created with LABEL='" 
             << label << "'" << std::endl;
        return oid._retn();
    } catch (const CF::Device::InvalidLocation & ) {
        throw; // rethrow
    } catch (...) {
        throw CF::Device::InvalidLocation();
    }
}

PortableServer::ObjectId * 
UserResourceServantProvider::create_application(const char* name)
    throw(CF::CreateApplicationError)
{
    try {
        std::string key(APP_PREFIX);
        key += name;
        ActiveObjectMap::iterator serv_pos = m_aom.find(key);
        if (serv_pos  != m_aom.end())
            throw CF::CreateApplicationError();
        PortableServer::ObjectId_var oid
            = PortableServer::string_to_ObjectId(key.c_str());

        ObjInfo info(APP,new Application_impl(m_object_deactivator,name));
        m_aom[key] = info;
        ECHO_CDMW << "APPLICATION <" << key << "> created with NAME='" 
             << name << "'" << std::endl;
        return oid._retn();
    } catch (const CF::CreateApplicationError & ) {
        throw; // rethrow
    } catch (...) {
        throw CF::CreateApplicationError();
    }
}

PortableServer::Servant 
UserResourceServantProvider::lookup_servant(const PortableServer::ObjectId& oid) 
    throw(PortableServer::ForwardRequest,
          CORBA::SystemException)
{
    try  {
        CORBA::String_var s = PortableServer::ObjectId_to_string(oid);
        std::string key(s.in());
        ActiveObjectMap::iterator serv_pos = m_aom.find(key);
        if (serv_pos  == m_aom.end()) {
            throw CORBA::OBJECT_NOT_EXIST(); // TODO Minor codes
        }
        else {
            ECHO_CDMW << "Activating " << key << " ... " << std::endl;
            return m_aom[key].m_serv;
        }
    } catch (const CORBA::BAD_PARAM & ) { // Pb with ObjectId
            throw CORBA::OBJECT_NOT_EXIST(); // TODO Minor codes
    } catch (const std::bad_alloc & ) { // failed to allocate memory
            throw CORBA::OBJECT_NOT_EXIST(); // TODO Minor codes
    }
}

void 
UserResourceServantProvider::release_servant(const PortableServer::ObjectId& oid,
                                             PortableServer::POA_ptr         poa,
                                             PortableServer::Servant         servant) 
    throw(CORBA::SystemException)
{
    
    Cdmw::LifeCycle::LifeCycleObjectBase * removable
        = dynamic_cast<Cdmw::LifeCycle::LifeCycleObjectBase *>(servant);
    try {
        CORBA::String_var s = PortableServer::ObjectId_to_string(oid);
        std::string key(s.in());
        ECHO_CDMW << "Deactivating " << key << " ... ";
        if ((removable) && removable->is_removed(oid)) {
            cleanup_object(oid);
            m_aom.erase(key);            
            servant->_remove_ref();
            ECHO_CDMW << " deleted.";
        }
//         m_aom.erase(key);            
//         servant->_remove_ref();
        ECHO_CDMW << std::endl;
    } catch (const std::bad_alloc & ) { // string alloc pb
            // TODO throw an exception (no memory)
    } catch (const CORBA::BAD_PARAM & ) { // A problem with oid
        // TODO throw an exception
    }
}

std::string
UserResourceServantProvider::get_supported_interface(const PortableServer::ObjectId& oid) const
    throw(CORBA::OBJECT_NOT_EXIST)
{
    try  {
        CORBA::String_var s = PortableServer::ObjectId_to_string(oid);
        std::string name(s.in());
        ActiveObjectMap::const_iterator serv_pos = m_aom.find(name);
        if (serv_pos  == m_aom.end()) {
            throw CORBA::OBJECT_NOT_EXIST(); // TODO Minor codes
        }
        else {
            std::string result("");
            if ((*serv_pos).second.m_type == DEV)
                result = DEV_REP_ID;
            else
                result = APP_REP_ID;
            return result;
        }
    } catch (const std::bad_alloc & ) { // Failed to allocate memory
            throw CORBA::OBJECT_NOT_EXIST(); // TODO Minor codes
    } catch (const CORBA::BAD_PARAM & ) { // Pb with ObjectId
            throw CORBA::OBJECT_NOT_EXIST(); // TODO Minor codes
    }
}
UserResourceServantProvider::ObjInfo::ObjInfo()
    throw()
    : m_type(0), m_serv(0)
{
}

UserResourceServantProvider::ObjInfo::ObjInfo(CF::ResourceNumType     type,
                                              PortableServer::Servant serv)
    throw()
    : m_type(type), m_serv(serv)
{
}
 
UserResourceServantProvider::ObjInfo::ObjInfo(const UserResourceServantProvider::ObjInfo & oi)
    throw()
    : m_type(oi.m_type), m_serv(oi.m_serv)
{
}
 
UserResourceServantProvider::ObjInfo & 
UserResourceServantProvider::ObjInfo::operator=(const UserResourceServantProvider::ObjInfo & rhs)
    throw()
{
    if (this != &rhs) {
        m_type = rhs.m_type;
        m_serv = rhs.m_serv;
    }
    return *this;
}

        
}; // End namespace Resources
}; // End namespace Cdmw

